game-core/src/main/java/games/strategy/engine/chat/HeadlessChat.java

```diff
@@ -73,11 +73,14 @@ public void addMessageWithSound(final String message, final PlayerName from, fin
         .start();
   }
 
-  private void addChatMessage(final String originalMessage, final String from) {
-    final String currentAllText = allText.toString();
-    if (currentAllText.length() > MAX_LENGTH) {
-      allText = new StringBuilder(currentAllText.substring(MAX_LENGTH / 2));
+  private void trimLengthIfNecessary() {
+    if (allText.length() > MAX_LENGTH) {
+      allText.delete(0, MAX_LENGTH / 2);
```

Comment:
Do you understand where the divide by 2 comes in? This one kinda confused me, if the length is over max, why are we dividing by 2 and not just truncating?

Reply:
I believe this "divide by 2" is to make sure that the string builder doesn't have to call delete again when the next message comes in, similar to how usually arraylist implementations double their allocated space whenever the underlying array gets full to maintain O(n) on average.

This shouldn't be a big deal now that the deletion operation isn't that expensive anymore, but I wouldn't bet on it